sky/skylet/providers/runpod/node_provider.py

```diff
@@ -0,0 +1,224 @@
+        config_tags.update(tags)
+        config_tags[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+
+        # Create nodes
+        ttype = node_config['InstanceType']
+        region = self.provider_config['region']
+
+        for _ in range(count):
+            instance_id = runpod_api.launch(name=self.cluster_name,
+                                            instance_type=ttype,
+                                            region=region)
+
+        if instance_id is None:
+            raise RunPodError('Failed to launch instance.')
+
+        runpod_api.set_tags(instance_id, config_tags)
+
+        instance_status = runpod_api.list_instances().get(instance_id, {})
+        while not (instance_status.get('status') == "RUNNING" and
+                   instance_status.get('ssh_port')):
+            time.sleep(3)
+            instance_status = runpod_api.list_instances().get(instance_id, {})
+
+        print(f"Instance {instance_id} is running and ready to use.")
+
+        return instance_id
+
+    @synchronized
+    def set_node_tags(self, node_id: str, tags: Dict[str, str]) -> None:
+        """Sets the tag values (string dict) for the specified node."""
+        node = self._get_node(node_id)
+        node['tags'].update(tags)
+        runpod_api.set_tags(node_id, node['tags'])
+        return None
+
+    def terminate_node(self, node_id: str) -> Optional[Dict[str, Any]]:
+        """Terminates the specified node."""
+        runpod_api.remove(node_id)
+        return None
+
+    @synchronized
+    def _get_filtered_nodes(self, tag_filters: Dict[str,
+                                                    str]) -> Dict[str, Any]:
+        """SkyPilot Method
+        Caches the nodes with the given tag_filters.
+        """
+        instances = runpod_api.list_instances()
+
+        filtered_nodes = {}
+        for instance_id, instance in instances.items():
+            if instance['status'] not in [
+                'CREATED', 'RUNNING', 'RESTARTING', 'PAUSED'
+            ]:
+                continue
+            if any(tag in instance['tags'] for tag in tag_filters):
+                filtered_nodes[instance_id] = instance
+
+        return filtered_nodes
+
+    def _get_node(self, node_id: str):
+        """ SkyPilot Method
+        Returns the node with the given node_id, if it exists.
+        """
+        instances = runpod_api.list_instances()
+        for instance_id, instance in instances.items():
+            if instance_id == node_id:
+                return instance
+
+        return None
+
+    def get_command_runner(
+        self,
+        log_prefix: str,
+        node_id: str,
+        auth_config: Dict[str, Any],
+        cluster_name: str,
+        process_runner: ModuleType,
+        use_internal_ip: bool,
+        docker_config: Optional[Dict[str, Any]] = None,
+    ) -> CommandRunnerInterface:
+        """Returns the CommandRunner class used to perform SSH commands.
+
```

Comment:
It would be nice to add a comment of why we are adding a customized command runner for runpod, such as `When external IP is used, runpod's default ssh port may not be 22, so we have to set the port used by the command runner.`